pkg/server/delta/v3/server.go

```diff
@@ -44,6 +132,340 @@ func NewServer(ctx context.Context, config cache.ConfigWatcher, callbacks Callba
+				}
+			case req.TypeUrl == resource.RouteType:
+				if values.deltaRouteNonce == "" || values.deltaRouteNonce == nonce {
+					if values.deltaRouteCancel != nil {
+						values.deltaRouteCancel()
+					}
+					values.deltaRoutes, values.deltaRouteCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.ListenerType:
+				if values.deltaListenerNonce == "" || values.deltaListenerNonce == nonce {
+					if values.deltaListenerCancel != nil {
+						values.deltaListenerCancel()
+					}
+					values.deltaListeners, values.deltaListenerCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.SecretType:
+				if values.deltaSecretNonce == "" || values.deltaSecretNonce == nonce {
+					if values.deltaSecretCancel != nil {
+						values.deltaSecretCancel()
+					}
+					values.deltaSecrets, values.deltaSecretCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			case req.TypeUrl == resource.RuntimeType:
+				if values.deltaRuntimeNonce == "" || values.deltaRuntimeNonce == nonce {
+					if values.deltaRuntimeCancel != nil {
+						values.deltaRuntimeCancel()
+					}
+					values.deltaRuntimes, values.deltaRuntimeCancel = s.cache.CreateDeltaWatch(req, &state)
+				}
+			default:
+				typeURL := req.TypeUrl
+				responseNonce, seen := values.deltaNonces[typeURL]
+				if !seen || responseNonce == nonce {
+					// We must signal goroutine termination to prevent a race between the cancel closing the watch
+					// and the producer closing the watch.
+					if terminate, exists := values.deltaTerminations[typeURL]; exists {
+						close(terminate)
+					}
+					if cancel, seen := values.deltaCancellations[typeURL]; seen && cancel != nil {
+						cancel()
+					}
+
+					var watch chan cache.DeltaResponse
+					watch, values.deltaCancellations[typeURL] = s.cache.CreateDeltaWatch(req, &state)
+
+					// a go-routine. Golang does not allow selecting over a dynamic set of channels.
+					terminate := make(chan struct{})
+					values.deltaTerminations[typeURL] = terminate
+					go func() {
+						select {
+						case resp, more := <-watch:
+							if more {
+								values.deltaResponses <- resp
+							} else {
+								// Check again if the watch is cancelled.
+								select {
+								case <-terminate: // do nothing
+								default:
+									// We cannot close the responses channel since it can be closed twice.
+									// Instead we send a fake error response.
+									values.deltaResponses <- deltaErrorResponse
+								}
+							}
+							break
+						case <-terminate:
+							break
+						}
+					}()
+				}
+			}
+		}
+	}
+}
+
 func (s *server) DeltaStreamHandler(str stream.DeltaStream, typeURL string) error {
-	return nil
+	// a channel for receiving incoming delta requests
+	reqCh := make(chan *discovery.DeltaDiscoveryRequest)
+	reqStop := int32(0)
+
+	go func() {
+		for {
+			req, err := str.Recv()
+			if atomic.LoadInt32(&reqStop) != 0 {
+				return
```

Comment:
Maybe use context cancelation here instead? Might be a bit cleaner

Reply:
I think that was previously tried but caused issues? I like that better too personally...

Reply:
What about something like this?
```go
func (s *server) DeltaStreamHandler(str stream.DeltaStream, typeURL string) error {
	// a channel for receiving incoming delta requests
	reqCh := make(chan *discovery.DeltaDiscoveryRequest)

	go func() {
		for {
			select {
			case <-str.Context().Done():
				close(reqCh)
				return
			default:
				req, err := str.Recv()
				if err != nil {
					close(reqCh)
					return
				}

				reqCh <- req
			}
		}
	}()

	return s.processDelta(str, reqCh, typeURL)
}
```

Reply:
This should utilize the stream context rather than try and track with the `reqStop` channel

Reply:
I just noticed they do this in SOTW:
```go
	// a channel for receiving incoming requests
	reqCh := make(chan *discovery.DiscoveryRequest)
	reqStop := int32(0)
	go func() {
		for {
			req, err := stream.Recv()
			if atomic.LoadInt32(&reqStop) != 0 {
				return
			}
			if err != nil {
				close(reqCh)
				return
			}
			select {
			case reqCh <- req:
			case <-s.ctx.Done():
				return
			}
		}
	}()

	err := s.process(stream, reqCh, typeURL)

	// prevents writing to a closed channel if send failed on blocked recv
	// TODO(kuat) figure out how to unblock recv through gRPC API
	atomic.StoreInt32(&reqStop, 1)

	return err
```
Maybe do that here?